### PR TITLE
Possibility to set exact integer values when editing unit traits in Map Editor

### PR DESCRIPTION
**Pull request by niklas-holma**
#18358 
A new Text Field Widget to show and edit traits values more exactly (integer resolution)

### CODE REVIEW

**Comment by phrohdoh** [comment]
I like it!

TS on 7fe052faa3ff30d6bd5f6dc3a68f25435df5329a
TS on 7fe052faa3ff30d6bd5f6dc3a68f25435df5329a

**Comment by pchote** [comment]
I'm not able to test this yet, but can someone confirm the behaviour when an invalid value (e.g. -1 or 200 for Health) is set?

**Comment by phrohdoh** [comment]
> I'm not able to test this yet, but can someone confirm the behaviour when an invalid value (e.g. -1 or 200 for Health) is set?

`-` is rejected completely, values (of the input and the actual init value) appear to be clamped on blur/focus lost

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: Does anyone have any thoughts on _requiring_ this text field on the actor props widget?

**Review by phrohdoh — DISMISSED** [review]
I think this will be useful and I haven't seen any obvious issues in the code or in testing.
Someone more familiar with widget and map editor code should take a look at this before merging.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (niklas-holma) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: I think all relevant yaml files have been modified. But yes, someone should double check that. It would be interesting to know if there are any other (valid) cases where the field would not exist.

Comment (phrohdoh) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: The field would not exist in 3rd-party games that haven't set it up (_may_ not want it, but I'm trying to determine that by asking for input), for the default games it is fine as you said.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: Style nit: no need for the explicit type or parens, so this could be `Action<float> updateValueField = f =>`

Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: But i'm not sure that the extra checks actually give us anything (I may be missing something), so this entire block could simplify down to something like:
```csharp
var valueField = sliderContainer.GetOrNull<TextFieldWidget>("VALUE");
if (valueField != null)
{
	Action<float> updateValueField = f => valueField.Text = ((int)f).ToString();
	updateValueField(so.GetValue(actor));
	slider.OnChange += updateValueField;

	valueField.OnTextEdited = () =>
	{
		float result;
		if (float.TryParse(valueField.Text, out result))
			slider.UpdateValue(result);
	};
}
```

Note that i've also changed `valueField` to use `GetOrNull` and added a null check to keep backwards compatibility - this is generally a good practice when working with non-critical UI additions.

Comment (niklas-holma) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: Right `((int)slider.Value) != ((int)result)` might be uneccessary, i think i added that as a precaution if the value had an initial float value (eg. 5.55), the value would be overwritten to an integer when you load the map/opened the popup.

The slider can use and produce float values (depending on its width), **but** i think the only fields that use it are integer: health and facing? I dont think they can be initialized to floating values ?

The changes look fine.

**Comment by dragunoff** [comment]
Works as advertised. It's a huge improvement over sliders-only.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: Can you update the PR, or would you prefer one of us to force push over the branch?

Comment (niklas-holma) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: I can update the PR

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (niklas-holma) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs]: I changed the code to use GetOrNull

File: OpenRA.Mods.Common/Widgets/SliderWidget.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/SliderWidget.cs:55]: Shouldn't this check `oldValue != Value`?

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:323]: This can use `out var`:
```
if (float.TryParse(valueField.Text, out var result))
```
(and dropping the `float result;` line.)

**Review by abcdefg30 — COMMENTED** [review]
Works fine ingame, this will be a nice addition.

**Comment by pchote** [comment]
I'd like to get this merged ASAP, so I filed #18652 with the fixups applied. I hope you don't mind!

**Pull request closed without merging**
